### PR TITLE
Fixes example4

### PR DESCRIPTION
**Pull request by agheata**
This does several unit-related fixes in example4, transportation+magnetic field, processes, and track. It fixes also some bugs in the field propagator.

### CODE REVIEW

**Comment by agheata** [comment]
Using these changes I'm getting the same number of hits and energy deposit for example4, but in general, there are still roundings that may make the transport diverge. This should be rebased, not squashed.

File: physics/processes/inc/transportation.h [hunk 1]
```diff
@@ -27,7 +27,7 @@ __host__ __device__ double transportation<fieldPropagator_t, BfieldOn>::transpor
 {
   // FIXME: Type should be double, but this leads to problems with below
   // condition checking 'step < physics_step'.
-  float step = 0.0;
+  double step = 0.0;
```
Comment (hahnjo) [inline physics/processes/inc/transportation.h:30]: If that works, please also remove the comment

File: magneticfield/inc/fieldPropagatorConstBz.h [hunk 2]
```diff
@@ -69,33 +69,33 @@ __host__ __device__ double fieldPropagatorConstBz::ComputeStepAndPropagatedState
     vecgeom::NavStateIndex &next_state)
 {
   double momentumMag = sqrt(kinE * (kinE + 2.0 * mass));
+  double momentumXYMag =
+      momentumMag * sqrt((1. - direction[2]) * (1. + direction[2])); // only XY component matters for the curvature
 
-  double curv   = std::fabs(ConstBzFieldStepper::kB2C * charge * BzValue) / (momentumMag + 1.0e-30); // norm for step
+  double curv = std::fabs(ConstBzFieldStepper::kB2C * charge * BzValue) / (momentumXYMag + 1.0e-30); // norm for step
```
Comment (jonapost) [inline magneticfield/inc/fieldPropagatorConstBz.h:75]: The safe length to progress in the plane will be proportional to momentumXYMag. But when we use it in 3-d it should become proportional to momentumMag.

File: magneticfield/inc/fieldPropagatorConstBz.h [hunk 3]
```diff
@@ -69,33 +69,33 @@ __host__ __device__ double fieldPropagatorConstBz::ComputeStepAndPropagatedState
...
+                                        // = sqrt( 2.0 / ( invEpsD * curv) ); // Candidate for fast inv-sqrt
 
   ConstBzFieldStepper helixBz(BzValue);
 
   double stepDone = 0.0;
-  double remains = physicsStep;
+  double remains  = physicsStep;
 
-  constexpr double epsilon_step = 1.0e-7; // Ignore remainder if < e_s * PhysicsStep
+  const double epsilon_step = 1.0e-7 * physicsStep; // Ignore remainder if < e_s * PhysicsStep
```
Comment (jonapost) [inline magneticfield/inc/fieldPropagatorConstBz.h:92]: Good idea!

File: magneticfield/inc/fieldPropagatorConstBz.h [hunk 4]
```diff
@@ -131,6 +131,7 @@ __host__ __device__ double fieldPropagatorConstBz::ComputeStepAndPropagatedState
       if (fullChord) {
         position  = endPosition;
         direction = endDirection;
+        move      = safeMove;
```
Comment (jonapost) [inline magneticfield/inc/fieldPropagatorConstBz.h:134]: Well caught!!

**Review by jonapost — APPROVED** [review]
I agree with key revisions.  
Small difference about safe length can be resolved in future.

**Comment by hahnjo** [comment]
For future reference: the precision changes of `eloss` in `energy_loss::GenerateInteraction` and of `esecond` in `pair_production::GenerateInteraction` give slightly different results for `example1` and `example2`.

**Pull request merged**